File: tests/udi_tests/thread.cxx
/*
 * Copyright (c) 2011-2015, UDI Contributors
 * All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

#include <cstdlib>
#include <cstring>
#include <iostream>
#include <sstream>
#include <set>

#include "libudi.h"
#include "libuditest.h"
#include "test_bins.h"
#include "test_lib.h"

using std::cout;
using std::endl;
using std::stringstream;
using std::string;
using std::set;
using std::pair;
using std::map;
using std::make_pair;

class test_thread : public UDITestCase {
    public:
        test_thread()
            : UDITestCase(std::string("test_thread")) {}
        virtual ~test_thread() {}

        bool operator()(void);
};

static const int NUM_THREADS = 10;

static const char *TEST_BINARY = THREAD_BINARY_PATH;
static udi_address TEST_FUNCTION = THREAD_BREAK_FUNC;

static test_thread testInstance;

/**
 * Used to process breakpoint events
 */
class BrkCallback : public EventCallback {

    public:
        BrkCallback(const set<udi_thread *> &threads)
            : threads(threads)
        {
            map<udi_thread *, udi_thread_state_e> thr_states;
            for (set<udi_thread *>::iterator i = threads.begin(); i != threads.end(); ++i) {
                thr_states.insert(make_pair(*i, UDI_TS_RUNNING));
            }
        }

        bool allEventsReceived() const { return threads.size() == 0; }

        virtual void operator()(udi_event *event);

    private:
        set<udi_thread *> threads;
        map<udi_thread *, udi_thread_state_e> thr_states;
};

void BrkCallback::operator()(udi_event *event) {
    test_assert(event->event_type == UDI_EVENT_BREAKPOINT);

    udi_event_breakpoint *brkpt = static_cast<udi_event_breakpoint *>(event->event_data);
    test_assert(brkpt->breakpoint_addr == THREAD_BREAK_FUNC);

    threads.erase(event->thr);

    validate_thread_state(thr_states);

    udi_error_e result = suspend_thread(event->thr);
    assert_no_error(event->proc, result);

    thr_states.insert(make_pair(event->thr, UDI_TS_SUSPENDED));

    validate_thread_state(thr_states);

    bool all_suspended = true;
    for (set<udi_thread *>::iterator i = threads.begin(); i != threads.end(); ++i) {
        if ( get_state(*i) == UDI_TS_RUNNING ) {
            all_suspended = false;
        }
    }

    if ( !all_suspended ) {
        result = continue_process(event->proc);
        assert_no_error(event->proc, result);
    }
}

/**
 * Used to process death events
 */
class DeathCallback : public EventCallback {
    public:
        DeathCallback(const set<udi_thread *> &threads)
            : threads(threads)
        {}

        bool allEventsReceived() const { return threads.size() == 0; }

        virtual void operator()(udi_event *event);

    private:
        set<udi_thread *> threads;
};

void DeathCallback::operator()(udi_event *event) {
    test_assert(event->event_type == UDI_EVENT_THREAD_DEATH);

    threads.erase(event->thr);

    udi_error_e result = continue_process(event->proc);
    assert_no_error(event->proc, result);
}

bool test_thread::operator()(void) {

    stringstream num_threads_str;
    num_threads_str << NUM_THREADS;

    string num_threads = num_threads_str.str();

    char *argv[] = { (char *)TEST_BINARY, (char *)num_threads.c_str(), NULL };

    udi_proc_config config;
    config.root_dir = NULL;
    
    udi_error_e error_code;
    char *errmsg = NULL;
    udi_process *proc = create_process(TEST_BINARY, argv, NULL, &config, &error_code, &errmsg);

    test_assert(proc != NULL);
    free(errmsg);
    test_assert(get_multithread_capable(proc));

    udi_thread *initial_thr = get_initial_thread(proc);
    test_assert(initial_thr != NULL);

    udi_error_e result = create_breakpoint(proc, TEST_FUNCTION);
    assert_no_error(proc, result);

    result = install_breakpoint(proc, TEST_FUNCTION);
    assert_no_error(proc, result);

    validate_thread_state(proc, UDI_TS_RUNNING);

    result = continue_process(proc);
    assert_no_error(proc, result);

    set<udi_thread *> threads;
    for (int i = 0; i < NUM_THREADS; ++i) {
        udi_thread *thr = wait_for_thread_create(proc);
        test_assert(thr != NULL);

        pair<set<udi_thread *>::iterator, bool> ins_result = threads.insert(thr);
        test_assert(ins_result.second);

        validate_thread_state(proc, UDI_TS_RUNNING);

        result = continue_process(proc);
        assert_no_error(proc, result);
    }

    wait_for_debuggee_pipe(proc);

    // tell debuggee to run threads into breakpoints
    release_debuggee_threads(proc);

    // Wait for all the breakpoints to occur
    BrkCallback brkCallback(threads);
    while (!brkCallback.allEventsReceived()) {
        wait_for_events(proc, brkCallback);
    }

    for (set<udi_thread *>::iterator i = threads.begin(); i != threads.end(); ++i) {
        resume_thread(*i);
    }

    result = continue_process(proc);
    assert_no_error(proc, result);

    // tell debuggee to let threads die
    release_debuggee_threads(proc);

    DeathCallback deathCallback(threads);
    while (!deathCallback.allEventsReceived()) {
        wait_for_events(proc, deathCallback);
    }

    wait_for_exit(initial_thr, EXIT_SUCCESS);

    cleanup_debuggee_pipe(proc);

    return true;
}

File: libudi-c/tests/thread.rs
//
// Copyright (c) 2011-2023, UDI Contributors
// All rights reserved.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
#![deny(warnings)]

use std::{ffi::CString, mem::MaybeUninit, path::PathBuf};

use udi_c::*;

mod native_file_tests;

const NUM_THREADS: u32 = 10;
const NUM_THREADS_STR: &str = "10";

#[test]
fn thread() -> Result<(), anyhow::Error> {
    let metadata = native_file_tests::get_test_metadata();
    let binary_path = metadata.workerthreads_path().to_str().unwrap();
    let thread_break_addr = metadata.thread_break_addr();
    let start_notification_addr = metadata.start_notification_addr();
    let term_notification_addr = metadata.term_notification_addr();

    unsafe {
        let executable = to_c_string(binary_path);
        let rt_lib_path = rt_lib_path();
        let config = udi_proc_config {
            root_dir: std::ptr::null(),
            rt_lib_path: to_c_string(rt_lib_path.as_ref().unwrap()),
        };
        let envp: &[*const libc::c_char] = &[std::ptr::null()];
        let argv: &[*const libc::c_char] = &[to_c_string(NUM_THREADS_STR), std::ptr::null()];
        let mut process: *const udi_process = std::ptr::null();

        to_result(create_process(
            executable,
            argv.as_ptr(),
            envp.as_ptr(),
            &config as *const udi_proc_config,
            &mut process,
        ))?;

        let mut initial_thr: *const udi_thread = std::ptr::null();
        to_result(get_initial_thread(process, &mut initial_thr))?;

        to_result(create_breakpoint(process, thread_break_addr))?;
        to_result(install_breakpoint(process, thread_break_addr))?;
        to_result(create_breakpoint(process, start_notification_addr))?;
        to_result(install_breakpoint(process, start_notification_addr))?;
        to_result(create_breakpoint(process, term_notification_addr))?;
        to_result(install_breakpoint(process, term_notification_addr))?;

        validate_thread_state(process, initial_thr, true)?;

        to_result(continue_process(process))?;

        let mut threads_created = 0;
        let mut start_received = false;

        // Wait for main thread to hit start_notification and suspend when it does
        // Wait for worker threads to be created
        while !start_received || threads_created < NUM_THREADS {
            let mut event_list: *const udi_event = std::ptr::null();

            to_result(wait_for_events(&process, 1, &mut event_list))?;

            let mut event = event_list;
            loop {
                match &(*event).event_type {
                    udi_event_type_e::UDI_EVENT_BREAKPOINT => {
                        let data = (*event).event_data as *const udi_event_breakpoint;
                        let addr = (*data).breakpoint_addr;
                        if addr == start_notification_addr {
                            start_received = true;

                            to_result(suspend_thread((*event).thr))?;
                        }
                    }
                    udi_event_type_e::UDI_EVENT_THREAD_CREATE => {
                        threads_created += 1;
                    }
                    _ => panic!("Unexpected event type"),
                }

                if (*event).next_event.is_null() {
                    break;
                }
                event = (*event).next_event;
            }

            if !event_list.is_null() {
                free_event_list(event_list);
            }

            if !start_received || threads_created < NUM_THREADS {
                to_result(continue_process(process))?;
            }
        }

        to_result(resume_thread(initial_thr))?;
        to_result(continue_process(process))?;

        let mut thread_breaks_received = 0;
        let mut term_received = false;

        while !term_received || thread_breaks_received < NUM_THREADS {
            let mut event_list: *const udi_event = std::ptr::null();

            to_result(wait_for_events(&process, 1, &mut event_list))?;

            let mut event = event_list;
            loop {
                match (*event).event_type {
                    udi_event_type_e::UDI_EVENT_BREAKPOINT => {
                        let data = (*event).event_data as *const udi_event_breakpoint;
                        let addr = (*data).breakpoint_addr;
                        if addr == term_notification_addr {
                            term_received = true;

                            to_result(suspend_thread((*event).thr))?;
                        } else if addr == thread_break_addr {
                            thread_breaks_received += 1;

                            to_result(suspend_thread((*event).thr))?;
                        } else {
                            panic!("Unexpected breakpoint event");
                        }
                    }
                    _ => panic!("Unexpected event type"),
                }

                if (*event).next_event.is_null() {
                    break;
                }
                event = (*event).next_event;
            }

            if !event_list.is_null() {
                free_event_list(event_list);
            }

            if !term_received || thread_breaks_received < NUM_THREADS {
                to_result(continue_process(process))?;
            }
        }

        validate_thread_state(process, initial_thr, false)?;

        let mut thr = initial_thr;
        while !thr.is_null() {
            to_result(resume_thread(thr))?;

            let mut next_thr = std::ptr::null();
            to_result(get_next_thread(process, thr, &mut next_thr))?;
            thr = next_thr;
        }

        to_result(continue_process(process))?;

        let mut thread_deaths_received = 0;

        while thread_deaths_received < NUM_THREADS {
            let mut event_list: *const udi_event = std::ptr::null();

            to_result(wait_for_events(&process, 1, &mut event_list))?;

            let mut event = event_list;
            loop {
                match (*event).event_type {
                    udi_event_type_e::UDI_EVENT_THREAD_DEATH => {
                        thread_deaths_received += 1;
                    }
                    _ => panic!("Unexpected event type"),
                }

                if (*event).next_event.is_null() {
                    break;
                }
                event = (*event).next_event;
            }

            if !event_list.is_null() {
                free_event_list(event_list);
            }

            if thread_deaths_received < NUM_THREADS {
                to_result(continue_process(process))?;
            }
        }

        to_result(continue_process(process))?;

        // Wait for exit event
        let mut event: *const udi_event = std::ptr::null();
        to_result(wait_for_events(&process, 1, &mut event))?;
        assert!((*event).next_event.is_null());
        match (*event).event_type {
            udi_event_type_e::UDI_EVENT_PROCESS_EXIT => {
                let data = (*event).event_data as *const udi_event_process_exit;
                assert_eq!(0, (*data).exit_code);
            }
            _ => panic!("Unexpected event type"),
        }
        free_event_list(event);

        to_result(continue_process(process))?;

        // Wait for process cleanup event
        let mut event: *const udi_event = std::ptr::null();
        to_result(wait_for_events(&process, 1, &mut event))?;
        assert!((*event).next_event.is_null());
        match (*event).event_type {
            udi_event_type_e::UDI_EVENT_PROCESS_CLEANUP => {}
            _ => panic!("Unexpected event type"),
        }
        free_event_list(event);

        Ok(())
    }
}

unsafe fn to_c_string(msg: &str) -> *const libc::c_char {
    let cstr = match CString::new(msg) {
        Ok(val) => val,
        Err(_) => return std::ptr::null(),
    };

    let len = cstr.to_bytes_with_nul().len();
    let output = libc::malloc(len);
    if !output.is_null() {
        libc::memcpy(output, cstr.as_ptr() as *const libc::c_void, len);
    }

    output as *const libc::c_char
}

unsafe fn to_result(result: udi_error) -> Result<(), anyhow::Error> {
    match result.code {
        udi_error_e::UDI_ERROR_NONE => Ok(()),
        udi_error_e::UDI_ERROR_NOMEM => Err(anyhow::anyhow!("Out of memory")),
        _ => {
            let msg = if result.msg.is_null() {
                "Unknown error".to_owned()
            } else {
                let msg = std::ffi::CStr::from_ptr(result.msg).to_str()?;
                let msg_str = msg.to_owned();
                libc::free(result.msg as *mut libc::c_void);
                msg_str
            };

            panic!("{}", msg);
            // Err(anyhow::anyhow!("{}", msg))
        }
    }
}

fn rt_lib_path() -> Option<String> {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("..");
    path.push("libudirt");
    path.push("build");
    path.push("src");
    path.push(sys::get_rt_lib_name());

    Some(path.to_str().unwrap().to_owned())
}

unsafe fn validate_thread_state(
    process: *const udi_process,
    initial_thr: *const udi_thread,
    running: bool,
) -> Result<(), anyhow::Error> {
    to_result(refresh_state(process))?;

    let mut thr = initial_thr;
    while !thr.is_null() {
        let mut state = MaybeUninit::uninit();
        to_result(get_state(thr, state.as_mut_ptr()))?;

        match state.assume_init() {
            udi_thread_state_e::UDI_TS_RUNNING => {
                if !running {
                    return Err(anyhow::anyhow!("Thread is running"));
                }
            }
            udi_thread_state_e::UDI_TS_SUSPENDED => {
                if running {
                    return Err(anyhow::anyhow!("Thread is suspended"));
                }
            }
        }

        let mut next_thr = std::ptr::null();
        to_result(get_next_thread(process, thr, &mut next_thr))?;
        thr = next_thr;
    }

    Ok(())
}

#[cfg(target_os = "linux")]
mod sys {
    pub fn get_rt_lib_name() -> &'static str {
        "libudirt.so"
    }
}

#[cfg(target_os = "macos")]
mod sys {
    pub fn get_rt_lib_name() -> &'static str {
        "libudirt.dylib"
    }
}

#[cfg(target_os = "windows")]
mod sys {
    pub fn get_rt_lib_name() -> &'static str {
        "udirt.dll"
    }
}
